l3build-arguments.lua

```diff
@@ -277,31 +276,31 @@ local function argparse()
     end
   end
   if next(names) then
-   result["names"] = names
+   result.names = names
   end
   return result
 end
 
-options = argparse()
-
 -- Sanity check
-function check_engines()
-  if options["engine"] and not options["force"] then
-     -- Make a lookup table
-     local t = { }
+A.check_engines = function (options, checkengines)
+  if options.engine and not options.force then
+    -- Make a lookup table
+    local t = {}
     for _, engine in pairs(checkengines) do
       t[engine] = true
     end
-    for _, engine in pairs(options["engine"]) do
+    for _, engine in pairs(options.engine) do
       if not t[engine] then
         print("\n! Error: Engine \"" .. engine .. "\" not set up for testing!")
         print("\n  Valid values are:")
         for _, engine in ipairs(checkengines) do
           print("  - " .. engine)
         end
         print("")
-        exit(1)
+        os.exit(1)
```

Comment:
Is there a  reason for this change? Standard Lua guides say to use local copies of _everything_ in production code.

Reply:
While I would keep the original code here for consistency reasons, it could be argued that specifically for `os.exit` not having a local copy is actually faster. (Since `os.exit` can never be called more than once, so the lookup is never duplicated)

Reply:
Eureka!
Now that makes more sense to me, this is not just a question of benchmark.
Then *everything* includes `print` and `checkengines` as well.
BTW, `check_engines` relies on `checkengines` and may be more appropriate somewhere else.

Reply:
> While I would keep the original code here for consistency reasons, it could be argued that specifically for `os.exit` not having a local copy is actually faster. (Since `os.exit` can never be called more than once, so the lookup is never duplicated)

Fair enough: I've not benchmarked, just went with a 'replace all of the globals' approach for standard stuff.

Reply:
> Then _everything_ includes `print` and `checkengines` as well.

It should definitely include `print`. `checkengines` is different because we explicitly *do not* want it to be fixed, we actually expect it to change after the file is loaded (the global will normally be nil when the file is first loaded)

In an abstract way, the distinction is between "fixed functions" (like `exit`, `print`, `next`) on the one hand which both should never change and even if they would change we probably would still want to call the original variant, while on the other hand there is "data" (like `arg`, `checkengines`, etc, but might also include callbacks which might technically be functions) which is much more variable and therefore more thought has to go into creating local copies for these.